### PR TITLE
Bug/swtch 885/improve csrf

### PR DESCRIPTION
**Pull request by hrachocode**

### CODE REVIEW

File: src/modules/auth/csurf.controller.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import { ApiExcludeEndpoint } from '@nestjs/swagger';
+import { Request, Response } from 'express';
+import { LoginGuard } from './login.guard';
+
+@Controller()
+export class CsurfController {
+
+    @UseGuards(LoginGuard)
+    @ApiExcludeEndpoint()
+    @Get('/get-frsc') // name should be unpredictable so hacker cannot grab it.
```
[the remark this answers is not in the document]
Comment (jrhender) [reply src/modules/auth/csurf.controller.ts:17]: Is the idea that this endpoint will be hit before login?

**Review by jrhender — COMMENTED** [review]
Looks like a good start. I'm not really understanding how the "double cookie submit" pattern will work in our case though. Will follow up with you.

File: .6e207494f2ab9f0e3cabc073301297a9300fc69f-audit.json [hunk 2]
```diff
@@ -0,0 +1,54 @@
+{
```
[the remark this answers is not in the document]
Comment (hrachocode) [reply .6e207494f2ab9f0e3cabc073301297a9300fc69f-audit.json:1]: its autogenerated :/ maybe some configs need to be changed?

File: src/modules/auth/csurf.controller.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (hrachocode) [reply src/modules/auth/csurf.controller.ts:17]: I think the endpoint should be called after login, as csrf protection is related to Claim POST requests which are requiring authorization as well.

didnt understand how being open-source is related to this:/

File: src/modules/auth/csrf.utils.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import csurf from "csurf";
+import { NextFunction, Request, Response } from "express";
+
+export const checkClientCsrf = (req: Request, res: Response, next: NextFunction) => {
+    if (!req.headers['server-request']) {
```
[the remark this answers is not in the document]
Comment (hrachocode) [reply src/modules/auth/csrf.utils.ts:5]: yeah, this is what I am thinking to optimize, I think we can store another server-secret-key and compare it with another token, kinda, not only checking header existence but also manual validation of server requests.

File: .6e207494f2ab9f0e3cabc073301297a9300fc69f-audit.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (hrachocode) [reply .6e207494f2ab9f0e3cabc073301297a9300fc69f-audit.json:1]: or maybe add this to envirement variable?

File: src/modules/auth/csrf.utils.ts [hunk 4]
```diff
@@ -11,7 +13,11 @@ export const checkClientCsrf = (req: Request, res: Response, next: NextFunction)
             }
         })
         csrf(req, res, next)
-    } else {
+    } else if (req.headers['server-request'] === process.env.CSRF_SERVER_SECRET) {
```
Comment (hrachocode) [inline src/modules/auth/csrf.utils.ts:16]: this way server requests will be validated by sending process.env.CSRF_SERVER_SECRET with headers when it's a server call, so hackers not only have to send server-request header but also have to guess the process.env.CSRF_SERVER_SECRET from .env file.

**Comment by jrhender** [comment]
@hrachocode I think this PR can be closed now, right?

**Pull request closed without merging**